signature/pki_cert.go

```diff
@@ -0,0 +1,100 @@
+		return errors.New("Internal inconsistency: PKI use set up without subject email or subject hostname")
+	}
+	return nil
+}
+
+func verifyPKI(pkiTrustRoot *pkiTrustRoot, untrustedCertificateBytes []byte, untrustedIntermediateChainBytes []byte) (crypto.PublicKey, error) {
+
+	untrustedLeafCerts, err := cryptoutils.UnmarshalCertificatesFromPEM(untrustedCertificateBytes)
+	if err != nil {
+		return nil, internal.NewInvalidSignatureError(fmt.Sprintf("parsing leaf certificate: %v", err))
+	}
+	switch len(untrustedLeafCerts) {
+	case 0:
+		return nil, internal.NewInvalidSignatureError("no certificate found in signature certificate data")
+	case 1:
+		break // OK
+	default:
+		return nil, internal.NewInvalidSignatureError("unexpected multiple certificates present in signature certificate data")
+	}
+	untrustedCertificate := untrustedLeafCerts[0]
+
+	if pkiTrustRoot.subjectEmail != "" {
+		if !slices.Contains(untrustedCertificate.EmailAddresses, pkiTrustRoot.subjectEmail) {
+			return nil, internal.NewInvalidSignatureError(fmt.Sprintf("Required email %q not found (got %q)",
+				pkiTrustRoot.subjectEmail,
+				untrustedCertificate.EmailAddresses))
+		}
+	}
+
+	if pkiTrustRoot.subjectHostname != "" {
+		if err = untrustedCertificate.VerifyHostname(pkiTrustRoot.subjectHostname); err != nil {
+			return nil, internal.NewInvalidSignatureError(fmt.Sprintf("Unexpected subject hostname: %v", err))
+		}
+	}
+
+	untrustedIntermediatePool := x509.NewCertPool()
+	if len(untrustedIntermediateChainBytes) > 0 {
+		untrustedIntermediateChain, err := cryptoutils.UnmarshalCertificatesFromPEM(untrustedIntermediateChainBytes)
+		if err != nil {
+			return nil, internal.NewInvalidSignatureError(fmt.Sprintf("loading certificate chain: %v", err))
+		}
+		if len(untrustedIntermediateChain) > 1 {
+			for _, untrustedIntermediateCert := range untrustedIntermediateChain {
+				if isIntermediateCA(untrustedIntermediateCert) {
+					untrustedIntermediatePool.AddCert(untrustedIntermediateCert)
+				}
+			}
+		}
+	}
+
+	if len(pkiTrustRoot.caIntermediateCertificates) > 0 {
+		trustedIntermediates, err := cryptoutils.UnmarshalCertificatesFromPEM(pkiTrustRoot.caIntermediateCertificates)
+		if err != nil {
+			return nil, internal.NewInvalidSignatureError(fmt.Sprintf("loading trusted intermediate certificates: %v", err))
+		}
+		for _, trustedIntermediateCert := range trustedIntermediates {
+			untrustedIntermediatePool.AddCert(trustedIntermediateCert)
+		}
+	}
+
+	if _, err := untrustedCertificate.Verify(x509.VerifyOptions{
+		Intermediates: untrustedIntermediatePool,
+		Roots:         pkiTrustRoot.caRootsCertificates,
+		KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageCodeSigning},
+	}); err != nil {
+		return nil, internal.NewInvalidSignatureError(fmt.Sprintf("veryfing leaf certificate failed: %v", err))
+	}
+
+	return untrustedCertificate.PublicKey, nil
+}
+
+// isIntermediateCA checks if the certificate is an intermediate CA
+func isIntermediateCA(cert *x509.Certificate) bool {
```

Comment:
It’s not immediately clear to me that this is necessary, I’d expect the `crypto/x509` code to validate the intermediates are usable. Is there a specific reason to filter them this way? If so, it would be nice to document that.

Reply:
I added this because I’m not sure if multiple root certificates will be included in the `untrustedIntermediateChainBytes` from `dev.sigstore.cosign/chain`. So, I’ve been filtering out intermediates from the data. Do you think this is necessary?

Reply:
By the principle of the thing, and reading `crypto/x509.Certificate.buildChains`, an intermediate entry is only trusted if it is signed by some other CA. So including any extra certificates should not hurt — after all, an attacker can include arbitrary intermediate certificates in the signature.

(Also — probably not for sigstore, but in general — there are transition scenarios, where a new CA might get its to-be-root certificate signed by some other CA; this to-be-root certificate can then be accepted as an intermediate.  That’s a valid situation (again, in general, probably not now for sigstore, and not for the BYOPKI users).

The certificate/issuer link is nowadays primarily happening via `SubjectKeyId`/`AuthorityKeyId` fields, so it seems to me it is _possible_ to for an intermediate to have `cert.Issuer == cert.Subject`.)